### ISSUE TITLE
Can CAF run on top of WebSocket connections?

### ISSUE THREAD

**Issue by xiewei20082008** [actor-framework/actor-framework#1888]
Now we need a feature that can enable the CAF protocol to run on top of WebSocket connections. I noticed the discussion at https://github.com/actor-framework/actor-framework/issues/1728, and also observed that the caf::net code is being frequently updated. I'd like to ask whether the CAF protocol over WebSocket is ready now? Or could you please provide some hints on the steps I need to take to implement it myself?

**Comment by Neverlord** [comment]
Do you really need the full-fledged CAF-to-CAF communication to run over WebSocket? What use case do you have?

If all you need is shipping (binary) CAF messages over WebSocket: you can serialize and deserialize a `caf::message` with the binary serializer and deserializer. If you need something that's more "stable", you can also implement a custom binary format (or use JSON over text frames). Maybe that's enough for your use case?

Porting the functionality from caf-io won't be a simple task. It's not just the protocol, you'll need to implement a middleman that establishes network-transparent messaging (and monitoring). We haven't tackled this yet, because we don't want to migrate the caf-io design as-is and there are still open questions on how we want to do addressing of actors in the future (cf. https://github.com/actor-framework/actor-framework/discussions/1421). Basically, caf-io had some limitations that we'd like to avoid with a redesign.

**Comment by xiewei20082008** [comment]
Initially, our product had a Node A and Node B that communicated internally. However, recently, one of our customers wanted to publish Node A on a public network to enable Node B to work remotely.

The issue is that their company's firewall policies only allow HTTPS traffic to the public network. Therefore, we would like to support this use case with minimal code changes.

**Comment by Neverlord** [comment]
Well, there are tools to tunnel TCP over WebSocket. Maybe that's the easiest way in your case?

**Comment by xiewei20082008** [comment]
Yes, that's what we're currently using as a temporary solution. However, this tunnel occasionally triggers antivirus software, which is quite annoying.
